[Jayanth-Balasubramanian] iP

Jayanth-Balasubramanian opened the pull request:
# DukePro

> "This chatbot is the greatest chatbot ever created, believe me. Nobody knows chatbots like I do, nobody. It's tremendous." - Former US President Donald Trump ([source](https://www.youtube.com/watch?v=r7l0Rq9E8MY))

Duke frees your mind of having to remember things to do. It's:

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use
All you need to do is:

1. Download it from the releases page.
2. Double click it.
3. Add your tasks.
4. Let it do the hard work of managing tasks for you 👍 .

And it is **FREE**!

Features:

- [x] Supports deadlines and events through the addition of `Deadline` and `Event` classes.
- [x] Supports sorting and filtering.

If you're a Java dev, you can learn quite a bit from reading the code! I have used the model-view-presenter and command patterns to structure the code. For example, here's a snippet from the `CommandFactory` class.
```java
public Command createCommand(CommandType type, String... args) throws InvalidParameterError {
        switch (type) {
        case GREET:
            return new GreetingCommand(taskView);
        case BYE:
            return new ByeCommand(taskView, exitEventListener);
        case LIST:
        ...
```

Commit:
Add Gradle support

Jayanth-Balasubramanian committed:
Initial commit: add basic functionality to chatbot

Jayanth-Balasubramanian committed:
Refactor to use MVP

Jayanth-Balasubramanian committed:
Add functionality to add tasks to list and display

Jayanth-Balasubramanian committed:
Add functionality to mark and unmark tasks as done

Jayanth-Balasubramanian committed:
Add todos, deadlines, events

Jayanth-Balasubramanian committed:
Add functionality to delete tasks

Jayanth-Balasubramanian committed:
Add error handling

Jayanth-Balasubramanian committed:
wrap all packages in duke package

Jayanth-Balasubramanian committed:
Add ability to save and load tasks from disk in TaskModel

Jayanth-Balasubramanian committed:
Make all dates LocalDateTime

Jayanth-Balasubramanian committed:
fix merge conflicts and merge branch-Level-8

Jayanth-Balasubramanian committed:
Add feature to find tasks

izzhafeez commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -1,23 +1,30 @@
-        this.deadline = deadline;
         this.taskView = taskView;
         this.taskModel = taskModel;
+        LocalTime defaultDeadlineTime = LocalTime.MAX;
+        this.deadline = DateTimeStringParser.parseDateTimeString(deadline, defaultDeadlineTime);
```
Hi, I'm your reviewer. I like the implementation of a failsafe for the deadline. Very elegant way to handle missing information.

Jayanth-Balasubramanian replied:
Thanks, appreciate the input!

izzhafeez reviewed (commented):
Looks good. Just the naming convention for commits needs a little standardization (capitalized vs uncapitalized).

ajjajjajjajj reviewed (commented):
Great work here! Here's some changes that you might want to consider implementing
- You might want to change names for `static final` variables to be in screaming snake case, and put them at the top of the declaration of class attributes - this helps to distinguish the constant variables 
- There's a wildcard import in `Duke.java`; minor violation
- A few cases of missing whitespace in your exception classes, before the first opening curly brace

Looks good otherwise though! Your code is really clear, and I learned quite a few things :)

chongweiguan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,39 @@
+
+import duke.exceptions.DukeException;
+import duke.exceptions.GlobalExceptionHandler;
+import duke.interfaces.*;
```
Hey! I'm one of your reviewers, and im supposed to look out for coding standard related issues! So here I see that you used a wildcard, you might want to list out the specific imports instead of using a wildcard import

chongweiguan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,39 @@
+        };
+
+        Presenter presenter = new TaskPresenter(taskModel, taskView, exitEventListener);
+        while(!duke.exit) {
```
minor violation here! need to add a space after while

chongweiguan commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,34 @@
+    private final String todoDescription;
+    private final LocalDateTime deadline;
+    private static final String addedTaskMessage = "Got it. I've added this task:\n  ";
+    AddDeadlineCommand(View taskView, TaskModel taskModel, String todoDescription, String deadline)
```
I think you have to add a space before this line? please correct me if im wrong though!

chongweiguan commented on src/main/java/duke/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,34 @@
+        this.deadline = DateTimeStringParser.parseDateTimeString(deadline, defaultDeadlineTime);
+    }
+
+
```
Minor issue: Delete one of the spaces here?

chongweiguan commented on src/main/java/duke/command/CommandFactory.java:
```diff
@@ -0,0 +1,65 @@
+                return new FindTaskCommand(taskModel, taskView, args[0]);
+            default:
+                // in case we add more to CommandType and forget to add here
+                return null;
```
Issue here! There is no indentation for case clauses

chongweiguan commented on src/main/java/duke/model/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.time.LocalDateTime;
+import duke.command.utils.DateTimeStringParser;
+public class Deadline extends Task {
```
Minor issue: Add a space before this line!

chongweiguan commented on src/main/java/duke/model/Event.java:
```diff
@@ -0,0 +1,26 @@
+package duke.model;
+import java.time.LocalDateTime;
+import duke.command.utils.DateTimeStringParser;
+public class Event extends Task{
```
Similar issue: add a line before this line

chongweiguan commented on src/main/java/duke/model/Event.java:
```diff
@@ -0,0 +1,26 @@
+    private final LocalDateTime startTime;
+    private final LocalDateTime endTime;
+
+
```
Minor issue: Delete a blank line from here!

chongweiguan commented on src/main/java/duke/presenter/InputParser.java:
```diff
@@ -0,0 +1,110 @@
+            throw new InvalidInputError(e.getMessage());
+        }
+
+//        String taskRegex = "^(.*?)(?: \\/by (.*)| \\/from (.*?) \\/to (.*))?$";
```
Minor issue here haha delete this line of comment!

chongweiguan reviewed (commented):
Overall coding standard is mostly followed! mostly minor issues, remember to change the case indentation! Great job and keep up the good work.

Jayanth-Balasubramanian committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Add gradle build support

Jayanth-Balasubramanian committed:
Fix: mark, unmark and delete tasks based on displayed list of tasks instead of list of all tasks

Jayanth-Balasubramanian committed:
Fix: gradle mainClassName so JAR file works

Jayanth-Balasubramanian committed:
Add checkstyle to gradle and checkstyle.xml

Jayanth-Balasubramanian committed:
Add javadocs to classes and (most) public methods

Jayanth-Balasubramanian committed:
Fix code standard violations

Jayanth-Balasubramanian committed:
Fix merge conflicts while merging A-CodingStandard and A-JavaDoc

Jayanth-Balasubramanian committed:
Delete binary tasks.ser file

Jayanth-Balasubramanian committed:
Add JUnit tests for TaskModel and TaskView

Jayanth-Balasubramanian committed:
Write tests to test public methods of TaskModel and TaskView

Jayanth-Balasubramanian committed:
Fix failing tests: update task model to return a copy of the task list

Jayanth-Balasubramanian committed:
Refactor view package into CLI and GUI packages

Jayanth-Balasubramanian committed:
Add simple GUI for Duke

Jayanth-Balasubramanian committed:
Modify other packages to work with GUI and keep them backwards compatible with the CLI

Jayanth-Balasubramanian committed:
Remove commented code from MainWindow

Jayanth-Balasubramanian committed:
Fix checkstyle errors, add JavaDocs and cite sources

Checkstyle errors were introduced during GUI development.
These were fixed, and JavaDocs were added or updated for the modified classes.

During development, StackOverflow snippets were used a few times without attribution,
these were fixed by adding a comment stating the author's name, date the answer was written,
and a link to the answer along with the name of the license under which the answer is provided.

Jayanth-Balasubramanian committed:
Fix licenses, add details about SE-EDU website licensing

Jayanth-Balasubramanian committed:
Use assertions to document assumptions in code

Important assumptions during code execution, such as list indices being less than the
length of the list and not being negative have been documented.

Jayanth-Balasubramanian committed:
Improve code quality by deleting TODO comments, renaming variables

Delete old TODO comments that were used in development. I renamed constant variables using snake case
and changed their access modifier to private static. I also added some comments explaining the main
Duke code which could potentially trip up someone reading the code base, such as the CommandEventListener
and InputEventListener.

Jayanth-Balasubramanian committed:
Merge pull request #2 from Jayanth-Balasubramanian/branch-A-Assertions

Use assertions to document assumptions in code

Jayanth-Balasubramanian committed:
Merge branch 'master' into branch-A-CodeQuality

Jayanth-Balasubramanian committed:
Merge pull request #3 from Jayanth-Balasubramanian/branch-A-CodeQuality

Improve code quality by deleting TODO comments, renaming variables

Jayanth-Balasubramanian committed:
Add deadline sorting functionality

Deadlines have a natural ordering, sorted by when they are due. When the user types "sort",
the SortDeadlinesCommand finds all the deadlines in the task list, and sorts them by when they are due,
then renders this to the view.

Jayanth-Balasubramanian committed:
Add a user guide

Jayanth-Balasubramanian committed:
Add screenshot of UI in docs/Ui.png

Jayanth-Balasubramanian committed:
Fix deadline sorting bug
